Framework/API/src/ExperimentInfo.cpp

```diff
@@ -1025,25 +1021,77 @@ ExperimentInfo::getInstrumentFilename(const std::string &instrumentName,
           to.setFromISO8601("2100-01-01T00:00:00");
 
         if (from <= d && d <= to) {
-          if (from > refDateGoodFile) { // We'd found a matching file more
-                                        // recently starting than any other
-                                        // matching file found
-            foundGoodFile = true;
-            refDateGoodFile = from;
-            mostRecentInstFile = pathName;
-          }
+          foundFile = true;
+          matchingFiles.insert(
+              std::pair<DateAndTime, std::string>(from, pathName));
         }
-        if (!foundGoodFile && (from >= refDate)) { // Use most recently starting
-                                                   // file, in case we don't
-                                                   // find a matching file.
+        // Consider the most recent file in the absence of matching files
+        if (!foundFile && (from >= refDate)) {
           refDate = from;
-          mostRecentInstFile = pathName;
+          mostRecentFile = pathName;
         }
       }
     }
   }
-  g_log.debug() << "Instrument file selected is " << mostRecentInstFile << '\n';
-  return mostRecentInstFile;
+
+  // Retrieve the file names only
+  std::vector<std::string> pathNames;
+  if (matchingFiles.size() > 0) {
```

Comment:
The emptyness of a container should be checked using `empty()`. While at it, I propose small modifications to the surrounding code as well:
```c++
  if (!matchingFiles.empty()) {
    pathNames.reserve(matchingFiles.size());
    for (auto elem : matchingFiles)
      pathNames.emplace_back(std::move(elem.second));
  } else {
    pathNames.emplace_back(std::move(mostRecentFile));
  }
```

Reply:
https://github.com/mantidproject/mantid/pull/24798/commits/ff6ff6f5c01f36a6c6dc68cd71535b48bf4a6539